src/plugins/esql_datagrid/public/data_grid.tsx

```diff
@@ -0,0 +1,143 @@
+import type {
+  DatatableColumn,
+  DatatableColumnMeta,
+  DatatableRow,
+} from '@kbn/expressions-plugin/common';
+import type { AggregateQuery } from '@kbn/es-query';
+import type { DataTableRecord } from '@kbn/discover-utils/types';
+import type { DataView } from '@kbn/data-views-plugin/common';
+import type { CoreStart } from '@kbn/core/public';
+import type { DataPublicPluginStart } from '@kbn/data-plugin/public';
+import type { FieldFormatsStart } from '@kbn/field-formats-plugin/public';
+import { RowViewer } from './row_viewer_lazy';
+
+interface ESQLDataGridProps {
+  core: CoreStart;
+  data: DataPublicPluginStart;
+  fieldFormats: FieldFormatsStart;
+  rows: ESQLRow[];
+  dataView: DataView;
+  columns: DatatableColumn[];
+  query: AggregateQuery;
+  flyoutType?: 'overlay' | 'push';
+  isTableView?: boolean;
+}
+type DataTableColumnsMeta = Record<
+  string,
+  {
+    type: DatatableColumnMeta['type'];
+    esType?: DatatableColumnMeta['esType'];
+  }
+>;
+
+const DataGrid: React.FC<ESQLDataGridProps> = (props) => {
+  const storage = new Storage(localStorage);
+  const [expandedDoc, setExpandedDoc] = useState<DataTableRecord | undefined>(undefined);
+  const [activeColumns, setActiveColumns] = useState<string[]>(
+    props.isTableView ? props.columns.map((c) => c.name) : []
+  );
+  const [rowHeight, setRowHeight] = useState<number>(5);
+
+  if (props.dataView.fields.getByName('@timestamp')?.type === 'date') {
+    props.dataView.timeFieldName = '@timestamp';
+  }
+
+  const renderDocumentView = useCallback(
+    (
+      hit: DataTableRecord,
+      displayedRows: DataTableRecord[],
+      displayedColumns: string[],
+      customColumnsMeta?: DataTableColumnsMeta
+    ) => (
+      <RowViewer
+        dataView={props.dataView}
+        toastNotifications={props.core.notifications}
+        hit={hit}
+        hits={displayedRows}
+        columns={displayedColumns}
+        columnsMeta={customColumnsMeta}
+        flyoutType={props.flyoutType ?? 'push'}
+        onRemoveColumn={(column) => {
+          setActiveColumns(activeColumns.filter((c) => c !== column));
+        }}
+        onAddColumn={(column) => {
+          setActiveColumns([...activeColumns, column]);
+        }}
+        onClose={() => setExpandedDoc(undefined)}
+        setExpandedDoc={setExpandedDoc}
+      />
+    ),
+    [activeColumns, props.core.notifications, props.dataView, props.flyoutType]
+  );
+
+  const columnsMeta = props.columns.reduce((acc, column) => {
+    acc[column.id] = {
+      type: column.meta?.type,
+      esType: column.meta?.esType ?? column.meta?.type,
+    };
+    return acc;
+  }, {} as DataTableColumnsMeta);
+
+  const columnNames = props.columns?.map(({ name }) => name);
+  const rows: DatatableRow[] = props.rows.map((row) => zipObject(columnNames, row));
+
+  return (
+    <UnifiedDataTable
+      columns={activeColumns}
+      rows={rows.map((row: Record<string, string>, idx: number) => {
```

Comment:
Also it would be great to use `useMemo`/`useCallback` for `rows`, `services`, `onSetColumns`, `sort` values so it does not recalculate on rerender.